source/common/upstream/wrsq_scheduler.h

```diff
@@ -0,0 +1,170 @@
+#pragma once
+
+#include <algorithm>
+#include <iostream>
+#include <memory>
+#include <queue>
+#include <utility>
+#include <vector>
+
+#include "envoy/common/random_generator.h"
+#include "envoy/upstream/scheduler.h"
+
+#include "common/common/assert.h"
+
+#include "absl/container/flat_hash_map.h"
+
+namespace Envoy {
+namespace Upstream {
+
+// Weighted Random Selection Queue (WRSQ) Scheduler
+// ------------------------------------------------
+// This scheduler keeps a queue for each unique weight among all objects inserted and adds the
+// objects to their respective queue based on weight. When performing a pick operation, a queue is
+// selected and an object is pulled. Each queue gets its own selection probability which is weighted
+// as the sum of all weights of objects contained within. Once a queue is picked, you can simply
+// pull from the top and honor the expected selection probability of each object.
+//
+// Adding an object will cause the scheduler to rebuild internal structures on the first pick that
+// follows. This operation will be linear on the number of unique weights among objects inserted.
+// Outside of this case, object picking is logarithmic with the number of unique weights. Adding
+// objects is always constant time.
+//
+// For the case where all object weights are the same, WRSQ behaves identical to vanilla
```

Comment:
Does wrsq handle first pick determinism for the case that all hosts have the same weight?

Reply:
This is something that will be addressed in a subsequent patch when plumbing WRSQ into the load balancer. In the case you mention, first pick determinism is fixed for WRSQ by simply adding hosts to the scheduler in a random order during refresh.